### ISSUE TITLE
LoRA / Parameter-Efficient Fine-Tuning

### ISSUE THREAD

**Issue by da2853** [starVLA/starVLA#229]
### Current State

StarVLA has **no LoRA/PEFT support**. Fine-tuning options are:
- Freeze entire VLM, train only action head (~300M params)
- Full fine-tuning of everything (~3.3B params)
- Custom adapter on action head only (`QwenAdapter` framework)

The adapter approach (`VLA_AdapterHeader.py`) adds learnable action queries (64 x 2048 = ~130K params) and cross-attention adapters, but does **not** adapt the VLM backbone itself.

### Module Freezing (current)

From `trainer_tools.py:151-193`:
```python
def freeze_backbones(model, freeze_modules=""):
    patterns = [p.strip() for p in freeze_modules.split(",")]
    for path in patterns:
        module = model
        for attr in path.split("."):
            module = getattr(module, attr)
        for param in module.parameters():
            param.requires_grad = False
```

Learning rate groups (`trainer_tools.py:40-107`) support per-module LRs:
- VLM backbone: 1e-5 (low)
- Action head: 1e-4 (10x higher)

### Adding LoRA

HuggingFace PEFT is compatible with Qwen2.5-VL. Implementation:

```python
from peft import get_peft_model, LoraConfig

lora_config = LoraConfig(
    r=16,
    lora_alpha=32,
    target_modules=["q_proj", "v_proj"],  # attention projections
    lora_dropout=0.05,
)
model.qwen_vl_interface.model = get_peft_model(
    model.qwen_vl_interface.model, lora_config
)
```

### Parameter Comparison

| Approach | Trainable Params | % of Total | Use Case |
|----------|-----------------|------------|----------|
| Full fine-tuning | ~3.3B | 100% | Pre-training, large data |
| Freeze VLM + train action head | ~300M | ~9% | Standard fine-tuning |
| Current adapter (action head only) | ~50-100M | ~1.5-3% | Lightweight adaptation |
| **LoRA r=16 on VLM + action head** | **~20-30M** | **~0.6-0.9%** | **Efficient fine-tuning** |
| **LoRA r=8 on VLM + action head** | **~10-15M** | **~0.3-0.5%** | **Minimal adaptation** |

### Potential Benefits

- **Composable**: Swap LoRA weights per robot/task without reloading base model
- **Memory efficient**: Only store/transmit delta weights
- **Fast training**: 10-50x fewer gradient computations
- **Community standard**: Most VLM fine-tuning uses LoRA now

### Files to Review

| File | Change |
|------|--------|
| `model/framework/base_framework.py` | Add `apply_lora()` method to base class |
| `training/trainer_utils/trainer_tools.py` | Add LoRA-aware parameter group construction |
| `config/training/*.yaml` | Add `lora:` config section |
| `requirements.txt` | Add `peft>=0.7.0` |

**Comment by nju-yqy** [comment]
Will the LoRA adaptation code be merged into the main branch? Besides its speed and efficiency, LoRA is quite important for methods that inject extra geometric info (like spatial forcing) to avoid corrupting the pre-trained backbone's distribution.

**Comment by JinhuiYE** [comment]
Thanks for the follow-up. LoRA or PEFT is still not supported on the current `starVLA_dev` branch. PR #371 is the current opt-in candidate, but it conflicts with the target branch and the joint-checkpoint contract still needs to be settled: when VLM LoRA and a fully trained action head are trained together, save and reload must preserve both, including the `save_adapter_only` case. Until the PR is rebased, validated through train-save-reload, and merged, please treat it as experimental rather than supported. We will keep this issue open and update it if support lands.
